subprojects/core/src/main/java/org/gradle/api/internal/DefaultDomainObjectCollection.java

```diff
@@ -103,21 +105,50 @@ protected DefaultDomainObjectCollection(DefaultDomainObjectCollection<? super T>
     }
 
     public Iterator<T> iterator() {
+        if (constantTimeIsEmpty()) {
+            return Iterators.emptyIterator();
+        }
         return new IteratorImpl(getStore().iterator());
     }
 
     public void all(Action<? super T> action) {
+
         action = whenObjectAdded(action);
 
+        if (constantTimeIsEmpty()) {
+            return;
+        }
+
         // copy in case any actions mutate the store
         // linked list because the underlying store may preserve order
-        Collection<T> copied = new LinkedList<T>(this);
-
-        for (T t : copied) {
-            action.execute(t);
+        // We make best effort not to create an intermediate collection if this container
+        // is empty.
+        Collection<T> copied = null;
+        for (T t : this) {
+            if (copied == null) {
+                copied = Lists.newArrayListWithExpectedSize(estimatedSize());
+            }
+            copied.add(t);
+        }
+        if (copied != null) {
+            for (T t : copied) {
+                action.execute(t);
+            }
         }
     }
 
+    /**
+     * Returns true if, and only if, the store is empty AND we know that we
+     * can query its size in constant time. Otherwise, it returns false, which
```

Comment:
Well, no. It means we don't know either it's empty or not empty in constant time. I'll rephrase.